File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/redis/RedisCachingConfiguration.java
package com.jmsoftware.maf.springcloudstarter.redis;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.cache.CacheManager;
import org.springframework.cache.annotation.CachingConfigurerSupport;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.cache.interceptor.*;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.data.redis.cache.RedisCacheConfiguration;
import org.springframework.data.redis.cache.RedisCacheManager;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.serializer.GenericJackson2JsonRedisSerializer;
import org.springframework.data.redis.serializer.RedisSerializationContext;

import java.util.Objects;

import static org.springframework.data.redis.cache.RedisCacheConfiguration.defaultCacheConfig;

/**
 * Description: RedisCachingConfiguration, change description here.
 *
 * @author 钟俊（zhongjun）, email: zhongjun@toguide.cn, date: 12/30/2020 11:48 AM
 * @see
 * <a href='https://blog.csdn.net/echizao1839/article/details/102660649'>Spring boot 之 spring-boot-starter-cache （整合 Redis）</a>
 **/
@Slf4j
@Configuration
@EnableCaching
@RequiredArgsConstructor
@ConditionalOnClass({RedisConnectionFactory.class})
public class RedisCachingConfiguration extends CachingConfigurerSupport {
    private final RedisConnectionFactory redisConnectionFactory;

    @Bean
    @Override
    public CacheManager cacheManager() {
        log.warn("Initial bean: '{}'", CacheManager.class.getSimpleName());
        RedisCacheConfiguration cacheConfiguration = defaultCacheConfig()
                .disableCachingNullValues()
                .serializeValuesWith(RedisSerializationContext.SerializationPair.fromSerializer(
                        new GenericJackson2JsonRedisSerializer()));
        return RedisCacheManager
                .builder(this.redisConnectionFactory)
                .cacheDefaults(cacheConfiguration)
                .build();
    }

    @Bean
    @Override
    public CacheResolver cacheResolver() {
        log.warn("Initial bean: '{}'", SimpleCacheResolver.class.getSimpleName());
        return new SimpleCacheResolver(Objects.requireNonNull(this.cacheManager()));
    }

    @Bean
    @Override
    public KeyGenerator keyGenerator() {
        log.warn("Initial bean: '{}'", KeyGenerator.class.getSimpleName());
        return (target, method, params) -> {
            val stringBuilder = new StringBuilder();
            stringBuilder.append(target.getClass().getName()).append("#").append(method.getName());
            for (Object param : params) {
                stringBuilder.append(":").append(param.toString());
            }
            return stringBuilder.toString();
        };
    }

    @Bean
    @Override
    public CacheErrorHandler errorHandler() {
        log.warn("Initial bean: '{}'", CacheErrorHandler.class.getSimpleName());
        return new SimpleCacheErrorHandler();
    }
}

File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/redis/RedisCachingConfiguration.kt
package com.jmsoftware.maf.springcloudstarter.redis

import com.jmsoftware.maf.common.util.logger
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.cache.CacheManager
import org.springframework.cache.annotation.CachingConfigurerSupport
import org.springframework.cache.annotation.EnableCaching
import org.springframework.cache.interceptor.*
import org.springframework.data.redis.cache.RedisCacheConfiguration
import org.springframework.data.redis.cache.RedisCacheManager
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.serializer.GenericJackson2JsonRedisSerializer
import org.springframework.data.redis.serializer.RedisSerializationContext
import java.lang.reflect.Method

/**
 * # RedisCachingConfiguration
 *
 * Description: RedisCachingConfiguration, change description here.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/15/22 9:08 PM
 * @see <a href='https://blog.csdn.net/echizao1839/article/details/102660649'>Spring boot 之 spring-boot-starter-cache （整合 Redis）</a>
 */
@EnableCaching
@ConditionalOnClass(RedisConnectionFactory::class)
class RedisCachingConfiguration(
    private val redisConnectionFactory: RedisConnectionFactory
) : CachingConfigurerSupport() {
    companion object {
        private val log = logger()
    }

    override fun cacheManager(): CacheManager {
        return RedisCacheManager
            .builder(redisConnectionFactory)
            .cacheDefaults(
                RedisCacheConfiguration.defaultCacheConfig()
                    .disableCachingNullValues()
                    .serializeValuesWith(
                        RedisSerializationContext.SerializationPair.fromSerializer(
                            GenericJackson2JsonRedisSerializer()
                        )
                    )
            )
            .build().apply {
                log.warn("Initial bean: `${CacheManager::class.java.simpleName}`")
            }
    }

    override fun cacheResolver(): CacheResolver {
        return SimpleCacheResolver(cacheManager()).apply {
            log.warn("Initial bean: `${SimpleCacheResolver::class.java.simpleName}`")
        }
    }

    override fun keyGenerator(): KeyGenerator {
        return KeyGenerator { target: Any, method: Method, params: Array<Any> ->
            val stringBuilder = StringBuilder()
            stringBuilder.append(target.javaClass.name).append("#").append(method.name)
            for (param in params) {
                stringBuilder.append(":").append(param.toString())
            }
            stringBuilder.toString()
        }.apply {
            log.warn("Initial bean: `${KeyGenerator::class.java.simpleName}`")
        }
    }

    override fun errorHandler(): CacheErrorHandler {
        return SimpleCacheErrorHandler().apply {
            log.warn("Initial bean: `${CacheErrorHandler::class.java.simpleName}`")
        }
    }
}
